### PR TITLE
Implement Grade Calculator (v1)

### LINKED ISSUE #245: Implement Standard Grade Calculator

### PR DESCRIPTION
**Pull request by rahulon12**
Fixes #245 

## Changes Made

- Implement a basic grade calculator, with the ability to create, add assignments and assignment groups. See score update live.
- Currently uses the groups and weights in the Canvas course, editable by the user.
- Reorder assignments and groups.
- Designs need refinement, but will be done iteratively over time.
- Future: Save edited user configuration of weights & assignments

## Screenshots (if applicable)

Simulator Screenshot - iPhone 16 Pro - 2025-03-01 at 13 43 01
Simulator Screenshot - iPhone 16 Pro - 2025-03-01 at 13 42 57
Screenshot 2025-03-01 at 1 41 57 PM
Screenshot 2025-03-01 at 1 41 44 PM

## Checklist
- [X] I have implemented all requirements for this PR and its accompanying issue.
- [X] I have verified my implementation across all edge cases.
- [X] I have ensured my changes compile on macOS & iOS.
- [X] I have resolved all SwiftLint warnings.

### CODE REVIEW

**Comment by rahulon12** [comment]
Please extensively test the calculator's math!

**Comment by rahulon12** [comment]
To do: displaying rules in the grade calc

**Comment by iwatger** [comment]
Screenshot 2025-03-01 at 8 03 28 PM
Can you make the sheet width adjustable? I have some grades that are 2+ digits in length, and i get a ... truncation with my current screen resolution/font size

**Comment by iwatger** [comment]
Screenshot 2025-03-01 at 8 16 48 PM

Is there a way to remove pre-existing or added assignments?

File: CanvasPlusPlayground/Features/Grades/GradeCalculator/GradeCalculator.swift [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+    init(assignmentGroups: [AssignmentGroup]) {
+        resetGroups(assignmentGroups)
+    }
+
+    // MARK: - User Intents
+    @discardableResult
+    func createEmptyGroup() -> GradeGroup {
+        let newGroup = GradeGroup(
+            id: UUID().uuidString,
+            name: "New Group",
```
Comment (iwatger) [inline CanvasPlusPlayground/Features/Grades/GradeCalculator/GradeCalculator.swift:118]: Can this be left as an empty string, so that when you start typing out the group name, you don't have to delete this text beforehand

**Comment by rahulon12** [comment]
> Screenshot 2025-03-01 at 8 16 48 PM
> Is there a way to remove pre-existing or added assignments?

Added. There was an issue with deleting groups, will add in a later iteration

**Comment by rahulon12** [comment]
> Screenshot 2025-03-01 at 8 03 28 PM
> Can you make the sheet width adjustable? I have some grades that are 2+ digits in length, and i get a ... truncation with my current screen resolution/font size

I've fixed the textfield width, although it's not as nicely aligned now.

**Comment by mko02** [comment]
https://github.com/user-attachments/assets/7bef73c3-2849-465f-bd97-25414640c7ba

Double clicking doesn't seem to allow editing, only single clicking works. My institution to edit is to double click not single click, but that may just be me

**Comment by mko02** [comment]
Screenshot 2025-03-02 at 4 47 44 PM

Minimizing the screen makes the pop up exceed the application

**Comment by mko02** [comment]
https://github.com/user-attachments/assets/5e892682-eb36-4953-ae52-d7d170ed32e8

Updating the percentage for the first time causes a glitch

**Comment by mko02** [comment]
https://github.com/user-attachments/assets/4a6c60ca-72c5-4055-bdeb-c06212e40572

adding more than 2 digits for the weight percentage causes the same glitch as above

**Comment by mko02** [comment]
Screenshot 2025-03-02 at 4 56 48 PM

The grade calculation right now is weighing each assignment based on the amount of points? I know some classes weigh each assignments equally, even though the `pointsPossible` is different. Could there be maybe an option to switch between the two formats?

**Comment by rahulon12** [comment]
> Screen.Recording.2025-03-02.at.4.45.22.PM.mov 
> 
> Double clicking doesn't seem to allow editing, only single clicking works. My institution to edit is to double click not single click, but that may just be me

I agree, but this is SwiftUI's default behavior :/

**Comment by rahulon12** [comment]
> Screen.Recording.2025-03-02.at.4.49.19.PM.mov 
> 
> Updating the percentage for the first time causes a glitch, the cursor should be before the %? There's no glitch if I edit it again. 1%2 is treated as 1%

This is a SwiftUI glitch when using the `percent` format for `TextField`. I just resorted to using a separate `Text` with "%", so it should be fixed now!

**Comment by rahulon12** [comment]
> Screenshot 2025-03-02 at 4 56 48 PM
> The grade calculation right now is weighing each assignment based on the amount of points? I know some classes weigh each assignments equally, even though the `pointsPossible` is different. Could there be maybe an option to switch between the two formats?

Yeah, we can do that in a future iteration. I'll create an issue for it. Right now the aim is to get the same "grade" value you get on Canvas

**Comment by rahulon12** [comment]
> Screenshot 2025-03-02 at 4 47 44 PM
> Minimizing the screen makes the pop up exceed the application

I'd consider this "expected" according to HIG

**Comment by azooz2003-bit** [comment]
@rahulon12 Why is the textfield selection so slow?

CleanShot 2025-03-05 at 22 27 43

**Comment by rahulon12** [comment]
> @rahulon12 Why is the textfield selection so slow?
> 
> CleanShot 2025-03-05 at 22 27 43 [ CleanShot 2025-03-05 at 22 27 43 ](https://private-user-images.githubusercontent.com/67667005/419738635-4c11d0d1-8d78-4ecf-adaa-8f3920ffae4f.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDEyMzU0NzMsIm5iZiI6MTc0MTIzNTE3MywicGF0aCI6Ii82NzY2NzAwNS80MTk3Mzg2MzUtNGMxMWQwZDEtOGQ3OC00ZWNmLWFkYWEtOGYzOTIwZmZhZTRmLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzA2VDA0MjYxM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBlZGQyNGM5YTEzODJlNWI5Y2ZkNTA1YTE3ZDMxNmNlNzY3ZTJiNjc3ZTVhMzI2YWQ4NTAxMTNlMDQwMDdhY2MmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.gJWaWFDQkaAEr-ZRorVKiYOPSJ3TBXb88MN3dOh7JrM) [ ](https://private-user-images.githubusercontent.com/67667005/419738635-4c11d0d1-8d78-4ecf-adaa-8f3920ffae4f.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDEyMzU0NzMsIm5iZiI6MTc0MTIzNTE3MywicGF0aCI6Ii82NzY2NzAwNS80MTk3Mzg2MzUtNGMxMWQwZDEtOGQ3OC00ZWNmLWFkYWEtOGYzOTIwZmZhZTRmLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzA2VDA0MjYxM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBlZGQyNGM5YTEzODJlNWI5Y2ZkNTA1YTE3ZDMxNmNlNzY3ZTJiNjc3ZTVhMzI2YWQ4NTAxMTNlMDQwMDdhY2MmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.gJWaWFDQkaAEr-ZRorVKiYOPSJ3TBXb88MN3dOh7JrM)

I looked into this, and it seems like a SwiftUI limitation when there are several text fields being rendered 😭

However, I believe there are ways to get around it, which will require refactoring. I will create a new umbrella issue with all the enhancements/perf issues that are mentioned by everyone in this PR. For the scope of this PR, I hope to focus on functionality and major bugs/code design.

**Comment by MJFree34** [comment]
I've been playing with it for a bit. A few things:
1. Shouldn't the current score on the main grades page be a percentage?
CleanShot 2025-03-11 at 01 48 57
2. What if we had the "Add Assignment" assignment have the same base points as others in the category? Like in this one, we could have it populate automatically with 100 points.
CleanShot 2025-03-11 at 01 50 21
3. It's a bit hard to realize how to delete assignments I create on mac. Instead of just swipe action we should also have it appear in the right click menu.
4. May just be me, but it's quite hard to see the icon for "Calculate Grades" button in much detail. I don't know if there is an SF Symbol that works besides the `+/-` symbol, but just a thought. Maybe ping the designers and see what they think?

I notoriously don't have like any grades except for that one course I played with so I don't have many schemes I was able to do with real data––just playing with the weightings of each course. It all seemed to work pretty solid with just the thoughts above!

**Comment by azooz2003-bit** [comment]
@rahulon12 I don't have any issues with the current implementation. Though I do think this needs a heavy redo (UI experience and grade calc choices) once designers come up with something.

File: CanvasPlusPlayground/Features/Grades/GradeCalculator/GradeCalculator.swift [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+            let totalEarned: Double = consideredAssignments.reduce(0.0) {
+                guard let pointsEarned = $1.pointsEarned else { return $0 }
+
+                return $0 + pointsEarned
+            }
+
+            guard totalPossible > 0 else { return nil }
+
+            return (totalEarned / totalPossible) * weight
+        }
```
Comment (azooz2003-bit) [inline CanvasPlusPlayground/Features/Grades/GradeCalculator/GradeCalculator.swift:98]: This is a bit of headache tbh haha. Can we have a document showing things like how you determined what goes into `consideredAssignments`?

File: CanvasPlusPlayground/Features/Grades/GradeCalculator/GradeCalculator.swift [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+                       let possible = assignment.pointsPossible {
+                        totalPoints += earned
+                        totalPossible += possible
+                    }
+                }
+            }
+
+            totalGrade = totalPossible > 0 ? (totalPoints / totalPossible) * 100 : 0.0
+        }
+    }
```
Comment (azooz2003-bit) [inline CanvasPlusPlayground/Features/Grades/GradeCalculator/GradeCalculator.swift:248]: same here, much easier if the calculation process is explained in a document

**Review by azooz2003-bit — APPROVED** [review]
im just gonna approve this so u can merge -  we need to move on - but plsss lets discuss how we want to approach this feature.
